### PR TITLE
feat: make binpack aware of worker kind

### LINKED ISSUE #230: Sampler accelerator sizing uses a flat 6 GiB KV reserve, so large models get placed on GPUs that are too small
### Summary

`src/server/estimator.py` sizes a sampler's accelerator budget as `params * 2 bytes + 6 GiB`. The 6 GiB reserve is meant to cover the KV cache, CUDA graphs and activations, but it does not scale with the model or the workload. For small models it is generous; for an 8B model it is already the bare minimum for the default GSM8K recipe batch, and it makes the scheduler prefer a 24 GB device over an 80 GB one.

### Observed

Qwen/Qwen3-8B, full fine-tune, `fft-gsm8k-rl` scenario defaults (`group_size=8`, `groups_per_batch=8`, `max_tokens=512`):

| Term | Value |
|---|---|
| weights, 8.19B params x 2 bytes | 16.4 GB |
| fixed sampler reserve, 6 GiB | 6.4 GB |
| sampler `accelerator_bytes` | 22.8 GB (21.25 GiB) |

Consequences on a fleet with L4 (24 GB) and H100 (80 GB) nodes:

1. **Tier choice.** The scheduler builds tiers ordered by least unused memory. The L4 tier ceiling is 23 GiB and the request is 21.25 GiB, so under `binpack` the 8B sampler is placed on an L4 even when H100 sampler nodes are idle. (Observed scheduler tiers for the request: `t1x23|t1x80`, L4 first.)
2. **vLLM budget on that L4.** `gpu_memory_utilization = min(budget / device, 0.90)` gives about 21.6 GB. After 16.4 GB of weights and 1-2 GB of runtime overhead, roughly 3-4 GB is left for KV cache. The recipe keeps 64 sequences in flight with ~700 tokens each (~45K tokens). Qwen3-8B costs ~144 KiB of KV per token (36 layers x 8 KV heads x 128 head_dim x 2 (K,V) x 2 bytes), so the batch alone needs ~6.5 GB. vLLM either fails startup profiling or runs with constant preemption.

Workaround used today: remove `openrl.io/sampler=true` from the L4 nodes so the only eligible tier is H100. That is a fleet-level hammer for a per-model sizing problem.

### Expected

The sampler budget should grow with the model so that a model whose KV cache cannot fit next to its weights on a small device is never sized to fit that device. Concretely, something like:

```
weights   = params * 2
kv_budget = kv_bytes_per_token(model) * nominal_tokens_in_flight   # from model config: layers, kv heads, head_dim
runtime   = 2 GiB                                                   # CUDA graphs, activations, allocator slack
device    = weights + max(kv_budget, floor) + runtime
```

where `kv_bytes_per_token` comes from the model config already loaded by `model_metadata.py`, and `nominal_tokens_in_flight` is a constant (or derived from the gateway's max concurrency). A simpler interim option is a percentage of the weights with a floor, for example `weights * 1.5 + 4 GiB`, which sizes the 8B sampler at ~29 GB and excludes the 24 GB tier naturally while still letting a 0.6B sampler share an L4.

Either way the reserve should be documented as a KV allowance, not a constant.

### Where

- `src/server/estimator.py`: `DEVICE_BYTES_PER_PARAM[("full", "sampler")] = 2`, `DEVICE_RESERVE_BYTES["sampler"] = 6 * GIB`, `footprint()`.
- `src/server/vllm_options.py`: `gpu_memory_utilization()` turns the budget into vLLM's fraction of the device.
- `scheduler/controller/internal/placement/tiers.go`: `Tiers()` orders candidate device sizes by least waste, which is what turns an under-estimate into a bad placement.

### Environment

open-rl main at 35f1e8a. GKE cluster with single-role pools: 2x L4 sampler nodes (2x24 GB each), 2x H100 sampler nodes (1x80 GB each). Placement strategy `binpack`.

### PR DESCRIPTION
**Pull request by ShubyM**
Because LoRA workers are not currently suspend-able they should be given exclusive access over a GPU and be torn down when done. This PR makes the scheduler aware of the kind of worker we are placing, FFT workers are suspendable so they can be timesliced against one another but LoRA cannot. Also changed the estimator to fix #230

### CODE REVIEW

File: scheduler/docs/design.md [hunk 1]
```diff
@@ -42,6 +42,13 @@ flowchart TD
     G --> H["node-local time-slicing"]
 ```
 
+Sharing is allowed only between workers whose `trainingKind` is `fft`, because
+those workers participate in suspension. LoRA and unspecified kinds receive
+exclusive claims under either strategy. Each ledger seat records its training
+kind; selection checks every occupant and booking repeats that check inside
+the ledger CAS loop. Old seats without a kind block new joins until their own
```
Comment (droot) [inline scheduler/docs/design.md]: why do we have to worry about the old seats ? (we can ignore backward compatibility if it is referring to that).

File: src/server/estimator.py [hunk 2]
```diff
@@ -23,7 +23,25 @@
...
+  "qwen3-1.7b": 114_688,  # 28 layers, 8 kv heads, head dim 128
+  "qwen3-4b": 147_456,  # 36 layers, 8 kv heads, head dim 128
+  "qwen2.5-7b": 57_344,  # 28 layers, 4 kv heads, head dim 128
+  "qwen3-8b": 147_456,  # 36 layers, 8 kv heads, head dim 128
+  "qwen3.5-9b": 32_768,  # 8 of 32 layers are full attention, 4 kv heads, head dim 256
+  "qwen3.5-27b": 65_536,  # 16 of 64 layers are full attention, 4 kv heads, head dim 256
+  "gemma-3-1b": 4_096,  # 4 of 26 layers are global, 1 kv head, head dim 256
+  "gemma-4-e2b": 7_168,  # 7 of 35 layers are full attention, 1 kv head, head dim 256
+  "gemma-4-e4b": 14_336,  # 7 of 42 layers are full attention, 2 kv heads, head dim 256
+}
```
Comment (droot) [inline src/server/estimator.py:43]: we might have to make this configurable. system will ship with the defaults, but will allow to extend the config through a configmap.

File: scheduler/controller/internal/controller/ledger.go [hunk 3]
```diff
@@ -85,18 +86,27 @@ func (r *WorkloadReconciler) ensureSeat(ctx context.Context, claimName string, s
...
+				return &claimLedger, existing, nil
 			}
-			// A predecessor's seat under our name: replace it. Its pod may
-			// still be terminating, and that is tolerated by two guards: the
-			// deterministic pod name is a mutex (the successor pod cannot be
-			// created until the old one is gone), and the fresh assignment ID
-			// fences the old incarnation from ever regaining residency.
-			*existing = seat
+			// Upgrade an old seat without changing its assignment or running pod.
+			existing.TrainingKind = seat.TrainingKind
```
Comment (droot) [inline scheduler/controller/internal/controller/ledger.go]: i wonder if we ignore backward compatibility, will that simplify this code.

File: scheduler/controller/api/v1alpha1/workload_types.go [hunk 4]
```diff
@@ -190,7 +190,7 @@ type Workload struct {
 	// The spec is immutable: every field either places the worker or renders
 	// its pod, and V1 does not re-place or re-render a live worker. Change by
 	// deleting and recreating.
-	// +kubebuilder:validation:XValidation:rule="self.role == oldSelf.role && self.modelID == oldSelf.modelID && has(self.ownerID) == has(oldSelf.ownerID) && (!has(self.ownerID) || self.ownerID == oldSelf.ownerID) && self.accelerator == oldSelf.accelerator",message="placement fields are immutable; delete and recreate the workload"
+	// +kubebuilder:validation:XValidation:rule="self.role == oldSelf.role && self.modelID == oldSelf.modelID && has(self.ownerID) == has(oldSelf.ownerID) && (!has(self.ownerID) || self.ownerID == oldSelf.ownerID) && has(self.trainingKind) == has(oldSelf.trainingKind) && (!has(self.trainingKind) || self.trainingKind == oldSelf.trainingKind) && self.accelerator == oldSelf.accelerator",message="placement fields are immutable; delete and recreate the workload"
 	Spec   WorkloadSpec   `json:"spec"`
```
Comment (droot) [inline scheduler/controller/api/v1alpha1/workload_types.go:202]: one thought i had:

we would also like to support workload that has exclusive access to accelerators irrespective of the the trainingKind. for ex. critical or latency sensitive workloads such as LAB job.

User can give the scheduling hint the session metadata "exclusive: true|yes"

one way to structure it will to have a field in the workload spec level to indicate that and we can set that field on the trainingKind or the metadata hint and drive the logic on the basis of that.

It might save time for our overall goal of productionization of LAB recipe.

**Review by droot — CHANGES_REQUESTED** [review]

File: scheduler/controller/api/v1alpha1/workload_types.go (hunk 4 above)
Comment (ShubyM) [reply scheduler/controller/api/v1alpha1/workload_types.go:202]: Adding a field called exclusive to workloadSpec, right now the gateway will make LoRA workers exclusive, later on we can add the wiring needed for people to pass metadata and solve (#200)

File: src/server/estimator.py (hunk 2 above)
Comment (ShubyM) [reply src/server/estimator.py:43]: Can do this in a separate PR

**Comment by ShubyM** [comment]
@droot PTAL

**Review by droot — APPROVED** [review]
